Allow configuration of behaviour when there is a Director Error in Python using define

blaisegassend opened the pull request:
Here is a less invasive version of my previous pull request. It introduces a SWIG_DIRECTOR_THROW macro into the Python director.swg that allows the behaviour of DirectorException to be customized. This approach is consistent with the existing SWIG_DIRECTOR_UEH.

ojwb commented:
There's no reason for this mechanism to be Python-specific, so the defaulting of `SWIG_DIRECTOR_THROW`should probably be done in a language-independent place, the testcase shouldn't be just for python, and appropriates uses of `throw` in `Lib/*/director.swg` should be changed to use the new macro.

Also, shouldn't the python code use a 4-space indent (pep8 and all that)?

blaisegassend commented:
Fixed the indentation.

For the location of the defaulting code, I was just following the model of
SWIG_DIRECTOR_UEH. Do you have any suggestion of what would be a good
common location?

I'm not sure I'm well enough versed in all the languages, but I can give a
stab at eliminating throw from the other languages and writing
corresponding tests. There seems to be a bit of a double standard here
though because a bunch of other features seem to only be implemented for a
subset of languages: SWIG_DIRECTOR_NORTTI, SWIG_DIRECTOR_STATIC are the
first two that I happened to look at.

On Wed, Mar 11, 2015 at 2:33 PM Olly Betts <notifications@github.com> wrote:

> There's no reason for this mechanism to be Python-specific, so the
> defaulting of SWIG_DIRECTOR_THROWshould probably be done in a
> language-independent place, the testcase shouldn't be just for python, and
> appropriates uses of throw in Lib/*/director.swg should be changed to use
> the new macro.
>
> Also, shouldn't the python code use a 4-space indent (pep8 and all that)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/swig/swig/pull/349#issuecomment-78377140>.
>

ojwb commented:
I believe the history of directors are that they were added for Python first, and then later added for other languages, one by one.  And each person who's subsequently added support for a language seems to have basically just copied the code for a language which already has director support and modified it for the new language.

As a result, there's both significant duplication of code, and significant amounts of inconsistency.  As an example, for no good reason I can see `SWIG_DIRECTOR_STATIC` is supported only for Python and PHP.  If it were supported for all languages then the fix for #353 would simply be to use `-DSWIG_DIRECTOR_STATIC`.  Meanwhile, `SWIG_DIRECTOR_UEH` is only supported for Python and Ruby, and `SWIG_DIRECTOR_RTTI` only for Python and Perl!

This makes it hard for users to understand directors as they lack commonality in unnecessary ways between languages, and it also means that we can end up having to fix a bug N times over.

So arguably I am applying a double standard, but only because I feel we need to start digging up rather than continuing to dig down.

There wasn't a director-specific common file, but it feels like we should have one, so In 7ba0652 I've created `Lib/director_common.swg` and moved the handling for `SWIG_DIRECTOR_STATIC` there.

blaisegassend commented:
OK, I can certainly move common looking code into Lib/director_common.swg.
Implementing tests for all the languages that use throw in director is a
lot more of a burden as I'm not well versed in all of them (I'm thinking of
the test writing). So is setting things up so that they can easily be added
in a consistent manner across languages without actually doing so in each
language an option?

On Wed, Mar 11, 2015 at 11:57 PM Olly Betts <notifications@github.com>
wrote:

> I believe the history of directors are that they were added for Python
> first, and then later added for other languages, one by one. And each
> person who's subsequently added support for a language seems to have
> basically just copied the code for a language which already has director
> support and modified it for the new language.
>
> As a result, there's both significant duplication of code, and significant
> amounts of inconsistency. As an example, for no good reason I can see
> SWIG_DIRECTOR_STATIC is supported only for Python and PHP. If it were
> supported for all languages then the fix for #353
> <https://github.com/swig/swig/issues/353> would simply be to use
> -DSWIG_DIRECTOR_STATIC. Meanwhile, SWIG_DIRECTOR_UEH is only supported
> for Python and Ruby, and SWIG_DIRECTOR_RTTI only for Python and Perl!
>
> This makes it hard for users to understand directors as they lack
> commonality in unnecessary ways between languages, and it also means that
> we can end up having to fix a bug N times over.
>
> So arguably I am applying a double standard, but only because I feel we
> need to start digging up rather than continuing to dig down.
>
> There wasn't a director-specific common file, but it feels like we should
> have one, so In 7ba0652
> <https://github.com/swig/swig/commit/7ba06526776274d779fb12636e1ab7e4063c0f72>
> I've created Lib/director_common.swg and moved the handling for
> SWIG_DIRECTOR_STATIC there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/swig/swig/pull/349#issuecomment-78432423>.
>

blaisegassend commented:
OK. I did the move into a common location, and I renamed the tests to be non-Python specific. All uses of throw have been replaced with SWIG_DIRECTOR_THROW.

I have not added tests to ocaml, perl5, php and ruby.

The test I have currently written for Python checks the functionality of SWIG_DIRECTOR_THROW. However, that does not test the most important thing: that the code should not raise any exceptions. 

I could do that relatively easily if I could #define throw to something else to cause its uses to break. Unfortunately, the code currently contains exception specifications, so there are some throw keywords that remain. Exception specifications appear to be generally regarded as a bad thing (http://stackoverflow.com/questions/1055387/throw-keyword-in-functions-signature). I'm curious why SWIG uses them? Would there be objections if I removed them from */director.swg? 

Alternately, would there be an easy way to build one of the tests with -fno-exceptions?

ojwb commented:
I don't think it's realistic to expect people to add a `runme` for every language when they add a new testcase, though I certainly wouldn't complain if they did.

Are you talking about the exception specifiers on methods of `DirectorException`?  AIUI, those need to be there as they are implementing virtual methods of `std::exception` which have the same exception specifiers (at least in C++98 - in C++11 these methods are marked `noexcept` instead - for example, the `what()` method - compare C++98 and C++11 tabs:

http://www.cplusplus.com/reference/exception/exception/what/

So it would probably be useful to replace those with a macro anyway, which defaults to `throw()` for C++98 and older and to `noexcept` for C++11 and newer.  If just removing the exception specifier works for your case, then you can pre-define it to be empty.

If that doesn't work, maybe you can `#ifdef` out `DirectorException` completely for your case?  If you aren't throwing exceptions in theory you don't need the exception classes, though perhaps it's not that simple to eliminate it.

ojwb commented:
> Alternately, would there be an easy way to build one of the tests with -fno-exceptions?

I think there's a mechanism for building a testcase with extra flags, but I've not tried to do it before so I'm not sure of the details.  I guess `-fno-exceptions` would need to be probed for, as it's not going to be available with all compilers.

blaisegassend commented:
You are quite right for the exception specifiers. And I haven't found any combination of preprocessor magic to do the test I wanted to do. That suggests that -fno-exceptions is the right way to go, but figuring out how to do that with automake seems to be more than a project than I have time to take on right now.

So what would you say about the patch as it currently stands?

ojwb commented on Examples/test-suite/python/director_throw_runme.py:
```diff
@@ -0,0 +1,29 @@
+    def go(self):
+      global go_called_count
+      go_called_count += 1
+      raise MyException
```
Looks like a non-standard indent here still.

blaisegassend replied:
Thanks for the autoconf snippet. It works except for a missing
AC_LANG_PUSH/AC_LANG_POP to switch to C++.

The exception specifications are not actually an issue even with
--fno-exceptions. On the other hand, playing with these tests made me
realize that there are autogenerated try-catch blocks that are a huge
problem. Going to have to look at those in more detail. -sigh-

Will also fix the indent.

On Thu, Mar 19, 2015 at 5:52 PM, Olly Betts <notifications@github.com>
wrote:

> In Examples/test-suite/python/director_throw_runme.py
> <https://github.com/swig/swig/pull/349#discussion_r26811930>:
>
> > @@ -0,0 +1,29 @@
> > +import director_throw
> > +
> > +class MyException(Exception):
> > +    pass
> > +
> > +go_called_count = 0
> > +class Derived(director_throw.Foo):
> > +    def go(self):
> > +      global go_called_count
> > +      go_called_count += 1
> > +      raise MyException
>
> Looks like a non-standard indent here still.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/swig/swig/pull/349/files#r26811930>.
>

ojwb commented:
Aside from the minor indent issue, the patch looks good to me.

But does it actually work for you with the exception specifiers there still?  Or is the issue with those just that you can't ensure that throw is never used by doing something like:

```
#define throw OOPS_THROW_SEEMS_TO_GET_USED
```

You'd want an autoconf test for `-fno-exceptions` (not automake) - something like this in configure.ac (untested):

```
AC_MSG_CHECKING([if -fno-exceptions works])
save_CXXFLAGS=$CXXFLAGS
CXXFLAGS="-fno-exceptions $CXXFLAGS"
FNO_EXCEPTIONS=
AC_COMPILE_IFELSE(
  [AC_LANG_PROGRAM([],[])],
  [AC_MSG_RESULT([yes])
  FNO_EXCEPTIONS=-fno-exceptions],
  [AC_MSG_RESULT([no])]
)
CXXFLAGS=$save_CXXFLAGS
AC_SUBST([FNO_EXCEPTIONS])
```

And then you can use `@FNO_EXCEPTION@` in generated files like `Makefile.in`.

blaisegassend commented:
The -fno-exceptions check was well worth writing as there were some catches getting autogenerated by the director module. Those are now wrapped in #ifdefs, and the director_no_stray_exceptions test allows me to see that the defines I have added are sufficient to turn off exceptions on all the platforms.

Assuming all the tests pass, this could be good to go.

blaisegassend commented:
OK, looks like making common.mk be autogenerated from common.mk.in broke the out-of-source testing. I'm part way down that rabbit-hole, but it won't be done this evening.

If you can see a better way of getting the @FNO_EXCEPTIONS@ option into common.mk where it is needed, I'm interested. I could set a variable in Examples/test-suite/*/Makefile, but that just feels like the wrong approach.

blaisegassend commented:
OK, I seem to have gotten the tests working (except possibly distclean), so this pull request should be in pretty good shape now assuming my latest fix for distclean works.

blaisegassend commented:
Is there an example of a passing AppVeyor test somewhere? AppVeyor seems brand new, and I don't see any other pull request that has actually had a run with it. So I'm not sure where to start to get things fixed.

ojwb commented:
Hmm, clicking on "Details" for the failing appveyor tests gives me a page which says:

"Project not found or access denied."

@wsfulton What's up with that?

But if it's really not passing for anything, it doesn't make much sense to block merges on it.

wsfulton commented:
I'd rather keep compiler specifics inside of '#ifdef __GNUC__' code in the actual testcase. If common.mk is changed, the dependencies need to be fixed so that the common.mk.in changes trigger a configure. This is likely to slow the build down though especially on Windows which suffers really badly from slow make.

Can you think of a way to put this into the actual .i file? I was was looking for a '#pragma GCC' but didn't succeed in getting it to work.

wsfulton commented:
It looks like the appveyor build is using a temporary appveyor project I set up which was subsequently deleted. My guess is that this pull request was created during the short period of time that that project was alive although I'd have expected the update to the pull request to run the new Appveyor project I've set up. More recent pull requests seem to be working on Appveyor.

blaisegassend commented:
Hi William,

Glad to see you back on the thread. It has come a long way under Olly's guidance so hopefully we'll manage to get this in without wasting as much of your time as if you had been involved throughout.

From your comments, I think the work that currently need so to happen is:
- Get AppVeyor working. I'm hoping that merging in master will go a long way here and will give it a shot.

- Figure out how to deal with -fno-exceptions. I really like having a test with -fno-exceptions because it actually enforces what I want to enforce and should be a good guard against regressions. I did some digging and so far haven't had more luck than you finding a way to enable -fno-exceptions other than the command-line flag. So the options I see at this point are:
  * Keep things similar to how I have them (with added checks to cause a reconfigure if common.mk.in is changed.
  * Have common.mk not be autogenerated by moving having a way to pass cxx flags into common.mk from each language's Makefile. I don't really like this option because it involves duplicating the rule that sets which targets need -fno-exceptions. And also I don't see how having common.mk being autogenerated differs significantly from having the various makefiles being autogerenated.
  * Alternate way of feeding -fno-exceptions into the compiler that you like better.
  * Find some magic that allows me to avoid using -fno-exceptions on the command line. I'm interested in any bright ideas here.

ojwb commented:
Instead of generating `config.mk` we could generate a small makefile fragment which just sets make variables for anything which is probed by `configure` which we need for testcases, then `include` that from `common.mk`.

blaisegassend commented:
Having a config.mk.in and leaving common.mk not autogenerated (I assume
that's what Olly meant) sounds like a reasonable idea to me. What do you
think, William?

On Tue, Apr 14, 2015 at 8:01 PM, Olly Betts <notifications@github.com>
wrote:

> Instead of generating config.mk we could generate a small makefile
> fragment which just sets make variables for anything which is probed by
> configure which we need for testcases, then include that from common.mk.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/swig/swig/pull/349#issuecomment-93167899>.
>

blaisegassend commented:
The merge seems to have helped the appveyor build. There is only one of the runs that is failing. It is failing by timing out with no log output at all. Is there some way to rerun the tests to see if this is a spurious failure without making a bogus commit and repushing? Not sure how to debug this given that there is no output.

ojwb commented:
Yes, that's what I was trying to suggest - I've clarified my previous comment.

blaisegassend commented:
William, would that make you happier?

On Wed, Apr 15, 2015 at 2:10 PM, Olly Betts <notifications@github.com>
wrote:

> Yes, that's what I was trying to suggest - I've clarified my previous
> comment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/swig/swig/pull/349#issuecomment-93570521>.
>

wsfulton commented:
I'd like to propose some changes and simplification.

The SWIG_DIRECTOR_CATCH_EXCEPTIONS macro addition ...
```
#ifdef SWIG_DIRECTOR_CATCH_EXCEPTIONS
  try {
#endif
...
#ifdef SWIG_DIRECTOR_CATCH_EXCEPTIONS
  } catch (Swig::DirectorException&) {
    SWIG_fail;
  }
#endif
```
could be done better via %feature so that the generated code can be
tweaked on a function by function basis and not just on the global level. This
will give more powerful / flexible control over the generated code. I 
was rather surprised to see the original implementation hard coded the try
catch block above and it does not get removed with %exception. So what I propose
is to add a flag attribute to %exception such that the try catch block can be
removed. Hence, the current code can be left as is by default, but the
following would generate the identical code, except for the addition of the
comment:
```
%exception %{
// %exception addition
$action
%}
```
which is the same as
```
%feature("except") %{
// %exception addition
$action
%}
```
If we add in the notrycatch attribute flag:
```
%feature("except", notrycatch=1) %{
// %exception addition
$action
%}
```
then for directors which previously generated the try catch block, the
try catch block not be generated.

That solves one part of the removal of try catch blocks. The 2nd part can use
%feature("director:except"). Use:
```
%feature("director:except") %{
// director:except tweak
if ($error) {
  abort();
}
%}
```
then this generated code:
```
    if (error) {
      Swig::DirectorMethodException::raise("Error detected when calling 'Manager.getPosition'");
    }
```
will be replaced by:
```
    // director:except tweak
    if (error) {
      abort();
    }
```
The above two features will thus control code generation on a function by function
basis or globally as %feature can if no method name is given. 

The third and final part is the globally generated code... the global
try catch blocks in the DirectorException's raise methods. As these are
global, a global macro is most appropriate and can be used to remove this
code:
```
#ifndef SWIG_DIRECTOR_NO_EXCEPTIONS
  /* base class for director exceptions */
  class DirectorException : public std::exception {
... all the classes derived from DirectorException ...
#endif
```
I havn't actually implemented this, but I think this covers all the cases you need to
rid the code of try, catch and throw. Let me know if it doesn't.

wsfulton commented:
I restarted the Appveyor build... looks okay now.

wsfulton commented:
A config.mk to sounds like a reasonable idea worth trying.

blaisegassend commented:
Hi William,

Let me go over what you said iterating through the macros I have added.

SWIG_DIRECTOR_RETHROW:
  I had a closer look and this one can go away completely as it was
actually only present in UEH.

SWIG_DIRECTOR_CATCH_EXCEPTIONS:
  The notrycatch flag is a great approach. Will implement.

SWIG_DIRECTOR_THROW:
  I'm not a big fan of the SWIG_DIRECTOR_NO_EXCEPTIONS flag that comments
out DirectorException, as it doesn't allow you to customize what you do per
type of exception. I also find it rather unpalatable that there are two
mechanisms for disabling the throws, %feature(director:except) which offers
you great flexibility and SWIG_DIRECTOR_NO_EXCEPTIONS which just hacks out
a bunch of code. I'd be more in favor of getting rid of all the raise()
methods in DirectorException, and throwing directly from the autogenerated
code (with a possible tweak to the exception constructor to handle any work
that was previously in throw). That way you can use director:except to
override all the throws and you don't need the macro. For example in this:

#ifdef __cplusplus
extern "C" {
#endif
SWIGINTERN PyObject *_wrap_TestClass_TestMethod(PyObject
*SWIGUNUSEDPARM(self), PyObject *args) {
  PyObject *resultobj = 0;
  TestClass *arg1 = (TestClass *) 0 ;
  void *argp1 = 0 ;
  int res1 = 0 ;
  PyObject * obj0 = 0 ;
  Swig::Director *director = 0;
  bool upcall = false;

  if (!PyArg_ParseTuple(args,(char *)"O:TestClass_TestMethod",&obj0))
SWIG_fail;
  res1 = SWIG_ConvertPtr(obj0, &argp1,SWIGTYPE_p_TestClass, 0 |  0 );
  if (!SWIG_IsOK(res1)) {
    SWIG_exception_fail(SWIG_ArgError(res1), "in method '"
"TestClass_TestMethod" "', argument " "1"" of type '" "TestClass *""'");
  }
  arg1 = reinterpret_cast< TestClass * >(argp1);
  director = SWIG_DIRECTOR_CAST(arg1);
  upcall = (director && (director->swig_get_self()==obj0));
#ifdef SWIG_DIRECTOR_CATCH_EXCEPTIONS
  try {
#endif
    if (upcall) {
      Swig::DirectorPureVirtualException::raise("TestClass::TestMethod");
    } else {
      (arg1)->TestMethod();
    }
#ifdef SWIG_DIRECTOR_CATCH_EXCEPTIONS
  } catch (Swig::DirectorException&) {
    SWIG_fail;
  }
#endif
  resultobj = SWIG_Py_Void();
  return resultobj;
fail:
  return NULL;
}

the raise line becomes:
      throw Swig::DirectorPureVirtualException("TestClass::TestMethod");

I think that goes in the same direction as what you were aiming for and
goes all the way.

What do you think?

Blaise


On Thu, Apr 16, 2015 at 11:45 AM, William S Fulton <notifications@github.com
> wrote:

> A config.mk to sounds like a reasonable idea worth trying.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/swig/swig/pull/349#issuecomment-93812039>.
>

wsfulton commented:
Can you please use the following in the testcase instead of modifying the build system:

```
/* Use recent versions of gcc to ensure there is no generated exception handling code (try catch or throw) */
#if __GNUC__ >= 5 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 7)
/* Equivalent to -fno-exceptions */
#pragma GCC optimize "no-exceptions"
#endif
```

wsfulton commented:
@blaisegassend - my thoughts were to replace all the additional macros you proposed with what I suggested... I'm not so keen on having lots of different macros to support, it is quite cumbersome. My thinking is if you don't want the default DirectorException family of classes, then one macro gets rid of them all, and you can simply slot in your own replacements. Otherwise it is quite hard to customise these exceptions with additional methods or code in the existing methods without a great big set of macros. I'm not sure I've really followed what you are proposing though, so maybe a patch is the next thing for me to look at what you think is best.

blaisegassend commented:
Awesome, I'll definitely bring in the pragma for no-exceptions. Makes the
whole patch orders of magnitude simpler!

For the DirectorException family of exceptions, since the exceptions are
actually being raised from autogenerated code, I'm proposing to
enable/disable them using a %feature just as you had suggested for the
other throw cases. It's going to be a fair bit of work to get it done in
all the languages so I'll probably do Python and ping you to see what you
think before moving on.

On Sat, Apr 18, 2015 at 3:19 AM, William S Fulton <notifications@github.com>
wrote:

> @blaisegassend <https://github.com/blaisegassend> - my thoughts were to
> replace all the additional macros you proposed with what I suggested... I'm
> not so keen on having lots of different macros to support, it is quite
> cumbersome. My thinking is if you don't want the default DirectorException
> family of classes, then one macro gets rid of them all, and you can simply
> slot in your own replacements. Otherwise it is quite hard to customise
> these exceptions with additional methods or code in the existing methods
> without a great big set of macros. I'm not sure I've really followed what
> you are proposing though, so maybe a patch is the next thing for me to look
> at what you think is best.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/swig/swig/pull/349#issuecomment-94151143>.
>

wsfulton commented:
hi @blaisegassend . There was quite a bit of work done in this patch and it would be a shame to waste it. Are you going to follow up with the next steps you mentioned?

blaisegassend commented:
Hi William,

I pretty much ran out of time. At this point I no longer have a use case,
and it has been long enough that it is completely context-switched out for
me. I don't expect to be able to find time to make the changes. I did hear
some interest from another googler a while back, but not sure if anything
will come of it.

Regards,
Blaise

On Sat, May 13, 2017 at 7:03 AM, William S Fulton <notifications@github.com>
wrote:

> hi @blaisegassend <https://github.com/blaisegassend> . There was quite a
> bit of work done in this patch and it would be a shame to waste it. Are you
> going to follow up with the next steps you mentioned?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swig/swig/pull/349#issuecomment-301250385>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB8TVXkliXcOTLtsa8xX7aBNerQ-6bLZks5r5bgygaJpZM4Dqifw>
> .
>



-- 

The information contained in this message is CONFIDENTIAL.  It is intended
only for the use of the individual or entity to whom it is addressed.  If
the reader of this message is not the intended recipient, you are hereby
notified that any dissemination or copying of this information is strictly
prohibited.  If you received this communication in error, please
immediately notify the sender, and thereafter destroy any electronic copies
of the message.

blaisegassend commented:
Yeah, we can close. Google has long since shifted to using CLIF for creating Python bindings.
